# Production Optimization  
Jiayin Ma |  April 2018
<br>

## Problem: 
You operate two plants, i.e., A and B. Each plant makes two products, “standard” and “deluxe”. A unit of standard gives a profit contribution of $10, while a unit of deluxe gives a profit contribution of $15.
Each plant uses two processes, grinding and polishing, for producing its products. Plant A has a grinding capacity of 80 hours per week and polishing capacity of 60 hours per week. For plant B, these capacities are 60 and 75 hours per week, respectively.
The grinding and polishing times in hours for a unit of each type of product in each factory are given in the tables below.  

<img width="709" alt="plants" src="https://user-images.githubusercontent.com/33208955/39788362-faef6cd8-52de-11e8-8cdd-50fa31189e32.png">  

It is possible, for example, that plant B has older machines than plant A, resulting in higher unit processing times. In addition, each unit of each product uses 4 kg of a raw material, which we refer to as raw. The company has 120 kg of raw available per week. To start with, we will assume that plant A is allocated 75 kg of raw per week and plant B the remaining 45 kg per week. Each plant can build a very simple linear programming model to maximize its profit contribution.

### 1. Optimization Problem
Find the optimal production allocation to maximize the revenue under the limits of capacity and raw materials.  

### 2. Decision variables
Let $i$ be the index of the plants:  
$i=1$ is plant A  
$i=2$ is plant B

Let $j$ be the index of the product type:  
$j=1$ is standard product  
$j=2$ is deluxe product  

let $k$ be the index of the process type:  
$k=1$ is grinding    
$k=2$ is polishing   

For example, $x_{11}$represents the number of standard products producing in plant A.  

### 3. Objective
\begin{equation}
Objective=10\times x_{i1} {+} {15}\times x_{i2}\
\end{equation}

### 4. Constraints  
Let $h_{ijk}$ be the process times in hours for a unit of each type of product in each factory. The capacity constraints for each factory are below:  

\begin{align}
x_{11}\times h_{111}+x_{12}\times h_{111}&\leq 80\\
x_{11}\times h_{112}+x_{12}\times h_{112}&\leq 60\\
x_{21}\times h_{211}+x_{21}\times h_{211}&\leq 60\\
x_{21}\times h_{222}+x_{22}\times h_{222}&\leq 75\\
\end{align}  

The raw materials constraints are:  

\begin{align}
(x_{11}+x_{12})\times {4}&\leq 75\\
(x_{21}+x_{22})\times {4}&\leq 45\\
\end{align}  
<br>

Assume that we could optimize the maerial allocation when we plan plant A and plant B together, the raw materials constraint would be:  

\begin{align}
(x_{11}+x_{12})\times {4}+(x_{11}+x_{12})\times {4}&\leq 120\\
\end{align}  

Processing capacities cannot be allocated since they are from two different plants/machines.  
<br>  

Non negative constraints:  
\begin{align}
x_{ij}&\geq 0\\
\end{align}  


### 5. Modeling

In [2]:
import cvxpy as cvx
import math
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [8]:
# define the variables
x=cvx.Bool(16)  # x=1 is when the hotel is selected to buy; x=0 is when the hotal is not selected to buy

price=np.matrix([2925000,10000000,3750000,3500000,325000,8950000,1950000,1750000,4900000,1650000,1125000,2500000,
                 1975000,3750000,1475000,750000])
profitability=np.matrix([44.24236879,53.37919231,43.02117894,42.6068584,37.34498761,49.09506947,23.77686566,
                         23.44540924,28.66584798,38.88067311,38.01059999,40.28936293,39.41928981,42.36096561,
                         38.59064874,37.38911919])

# objective
objective=cvx.Maximize(cvx.sum_entries(profitability*x))  # to maximize the average predicted profitability

#x_sum=sum(x[0]+x[1]+x[2]+x[3]+x[4]+x[5]+x[6]+x[7]+x[8]+x[9]+x[10]+x[11]+x[12]+x[13]+x[14]+x[15])  # number of hotels purchased

# contraints
c1=cvx.sum_entries(price*x)<=10000000  # $10 million budget contraint
con=[c1]

prob=cvx.Problem(objective, con)
result=prob.solve()
print ("The optimal profitability is: ", result)
print ("The hotels we should purchase are: ", x.value)

The optimal profit for plant A is:  220.00000005215267
Optimal allocation of standard products:  10.000000000492811
Optimal allocation of deluxe products 8.000000003148305
------------------------------------------------------------------------------------
The optimal profit for plant B is:  165.00000006681978
Optimal allocation of standard products:  2.912859137415921e-09
Optimal allocation of deluxe products:  11.000000002512746


In [10]:
# joint optimization of Plant A and B without changing the materials constraints

con=[c11,c12,c21,c22,c3,c4,c5,c6,c7,c8]  # constraints for both plant A and plant B

prob = cvx.Problem(objective, con)  # solution for both plant A and plant B
result = prob.solve()
print ("The optimal profit for plant A and plant B is: ", result)
print ("Optimal allocation of standard products in plant A: ", x11.value)
print ("Optimal allocation of deluxe products in plant A: ", x12.value)
print ("Optimal allocation of standard products in plant B", x21.value)
print ("Optimal allocation of deluxe products in plant B", x22.value)

The optimal profit for plant A and plant B is:  385.0000000611009
Optimal allocation of standard products in plant A:  10.000000002876392
Optimal allocation of deluxe products in plant A:  8.000000001475799
Optimal allocation of standard products in plant B 3.0780839580717643e-09
Optimal allocation of deluxe products in plant B 10.999999998627942


In [11]:
# Assume that we could optimize the material allocation
# joint optimization of Plant A and B by changing the materials constraints

c00=(x11+x12+x21+x22)*4<=120  # assume that the only constraint of raw materials is 120kg when consider plant A and B tpgether
con_r=[c11,c12,c21,c22,c5,c6,c7,c8,c00]  # assume that we can reallocate the raw materials

prob_r = cvx.Problem(objective, con_r)  # solution for plant A and plant B
result_r = prob_r.solve()
print ("The optimal profit for plant A and plant B is: ", result_r)
print ("Optimal allocation of standard products in plant A: ", x11.value)
print ("Optimal allocation of deluxe products in plant A: ", x12.value)
print ("Optimal allocation of standard products in plant B", x21.value)
print ("Optimal allocation of deluxe products in plant B", x22.value)

The optimal profit for plant A and plant B is:  399.99999992548265
Optimal allocation of standard products in plant A:  9.999999999202235
Optimal allocation of deluxe products in plant A:  7.999999996534886
Optimal allocation of standard products in plant B -1.4098916005133095e-09
Optimal allocation of deluxe products in plant B 11.999999999969063


### 6. Summary  
* When we optimoze the profit for the two plants separately, the total optimal profit is about 385. When we optimoze the profit for the two plants at the same time, the total optimal profit is about 400. **Therefore, we can achieve higher total profit by the joint optimization for plant A and plant B than by the separaterd optimization.**  

* The only difference between the two optimization model is the raw material constraint. In the separated model, plant A is allocated 75kg raw material while plant B is allocated 45kg material. In the joint model, plant A and plant B are allocated 120kg totally. **The CEO of the company should improve the raw material allcation for each plant to make the best of the material so that the company can achieve higher profit.**  

* The optimal solution is to produce 10 standard products ans 8 deluxe prodects in plant A while producing 12 deluxe products in plant B.  
<br>